### DISCUSSION TITLE
How to set request specific data(Variables) with `app.fetch()`?

### DISCUSSION THREAD

**Discussion by MellKam** [honojs/hono#3615] [Q&A]
I'm using Hono as a router and RPC client in my Astro app. I'm trying to figure out if there's a way to set variables with the .fetch function. I have middleware on the Astro side handling authorization, so I want to pass user and session objects to Hono when making a request. But it looks like fetch only takes bindings as the second parameter.

```ts
export const app = new Hono<{
	Variables: {
		session: Session | null;
		user: User | null;
	};
}>()
	.basePath("/api")
	.route("/auth", authRouter)
	.route("/categories", categoriesRouter);
// ...

export const ALL: APIRoute = (ctx) => {
	return app.fetch(ctx.request, { user: ctx.locals.user, session: ctx.locals.session });
};
```
I tried passing `user` and `session` as Bindings, which works, but they don’t seem ideal for data specific to each request. The ultimate fix might be to create Hono middleware and skip using Astro locals, which I’ll probably do, but I’d still like to know if there’s a way to achieve this directly.

**Comment by StephenHaney** [comment]
Just ran into the same question. I'm using `Bun.serve` in front of Hono and handling my session refresh before the Hono app. I'd like to pass the validated user into Hono's `app.fetch`. It looks like you can do it with Bindings, which is what I'm using. But I believe `variables` would be more what Hono intends here.

**Comment by manterfield** [comment]
Nothing to add on the description of the problem. Just vouching for the use case: I'm using the same setup (astro passing the request onto Hono) and bumped into the same thing.

I'm using Astro in cloudflare pages and set it up this way to keep my API in the same worker. There's a stack of 'stuff' I'd like to pass on and bindings feels semantically odd as a place to store it.

**Comment by kylemclean** [comment]
Agreed, it would be nice to be able to put request-specific data in variables instead of bindings.

You can use Hono middleware to set the variables instead, but if you want the data to also be accessible from your framework code (e.g. server-side loader functions), this isn't really an option.
